Add support for computing the inverse hyperbolic cotangent (acoth)

BrunoFenzl opened the pull request:
Resolves #236.

<!--lint disable first-heading-level-->

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [x] Read and understand the [Code of Conduct][code-of-conduct].
-   [x] Read and understood the [licensing terms][license].
-   [x] Searched for existing issues and pull requests **before** submitting this pull request.
-   [x] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [x] Rebased onto latest `develop`.
-   [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

-   implements issue #236 

## Related Issues

> Does this pull request have any related issues?
no

This pull request:

-   resolves #236

## Questions

> Any questions for reviewers of this pull request?
tex notation docs are not up to date. I have to read my way through to update them.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Happy to have some feedback!

* * *

@stdlib-js/reviewers

<!-- <links> -->

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

<!-- </links> -->

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/README.md:
```diff
@@ -0,0 +1,87 @@
+
+# acoth
+
+> Compute the [inverse hyperbolic cotangent][inverse-hyperbolic-cotangent] of a number.
```
```suggestion
> Compute the [inverse hyperbolic cotangent][inverse-hyperbolic-cotangent].
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/README.md:
````diff
@@ -0,0 +1,87 @@
+
+```javascript
+var v = acoth( 1.0 );
+// returns 0
````
```suggestion
// returns 0.0
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/README.md:
````diff
@@ -0,0 +1,87 @@
+// returns NaN
+```
+
+The domain of `x` is restricted to `[x < -1 || x > 1)`. If `x >= -1 || <= 1`, the function will return `NaN`.
````
```suggestion
If `x` is on the interval `[-1, 1]`, the function returns `NaN`.
```

BrunoFenzl replied:
On a second look, if `x` is exactly -1 or 1, this implementation it will return 0. NaN will be returned if `x` is less than 1 and more than -1 but not equal. This was my mistake.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js:
```diff
@@ -0,0 +1,79 @@
+
+// VARIABLES //
+
+var opts = {
```
This can be removed, along with the "built-in" benchmark. The built-in `Math` object does not have a `acoth` method, nor are there any proposals to add to the language.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js:
```diff
@@ -0,0 +1,79 @@
+
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*100.0 ) + 1.0;
```
```suggestion
		x = ( randu()*100.0 ) + 1.1;
```

`randu()` returns values on the interval `[0,1)`. In which case, need to add a value slightly greater than `1` in order to ensure that `y` is not `NaN`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/benchmark.js:
```diff
@@ -0,0 +1,79 @@
+	b.end();
+});
+
+bench( pkg+'::built-in', opts, function benchmark( b ) {
```
This benchmark can be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,134 @@
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( 100.0*rand_double() ) + 1.0;
+		y = acoth( x );
```
`acoth` is not a built-in C/C++ function. You'll need to actually provide an implementation, either as a separate function or inlined. For example,

```suggestion
y = atanh( 1.0/x );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/Makefile:
```diff
@@ -0,0 +1,113 @@
+#/
```
Cephes does not have an `acoth` function. You can remove this file.

kgryte replied:
This file still needs to be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/c/cephes/benchmark.c:
```diff
@@ -0,0 +1,139 @@
+/**
```
Cephes does not have an `acoth` function. You can remove this file.

kgryte replied:
This file still needs to be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/Makefile:
```diff
@@ -0,0 +1,110 @@
+#/
```
To my knowledge, Boost does not have an `acoth` function. You can remove this file.

kgryte replied:
This file still needs to be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/cpp/boost/benchmark.cpp:
```diff
@@ -0,0 +1,134 @@
+/**
```
To my knowledge, Boost does not have an `acoth` function. You can remove this file.

kgryte replied:
This file still needs to be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/julia/benchmark.jl:
````diff
@@ -0,0 +1,143 @@
+```
+"""
+function benchmark()
+	t = BenchmarkTools.@benchmark acoth( (100.0*rand()) + 1.0 ) samples=1e6
````
```suggestion
	t = BenchmarkTools.@benchmark acoth( (100.0*rand()) + 1.1 ) samples=1e6
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py:
```diff
@@ -0,0 +1,97 @@
+def benchmark():
+    """Run the benchmark and print benchmark results."""
+    setup = "from math import acoth; from random import random;"
+    stmt = "y = acoth(100.0*random() + 1.0)"
```
Python does not have an `acoth` function. You need to modify this as follows:

```suggestion
stmt = "y = atanh(1.0 / ((100.0*random() + 1.1))"
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/r/benchmark.R:
```diff
@@ -0,0 +1,109 @@
+	#' elapsed <- benchmark( 10000L );
+	benchmark <- function( iterations ) {
+		# Run the benchmarks:
+		results <- microbenchmark::microbenchmark( acoth( (100.0*runif(1)) + 1.0 ), times = iterations );
```
To my knowledge, R does not have an `acoth` function. You need to modify this to

```suggestion
results <- microbenchmark::microbenchmark( atanh( 1.0 / ( (100.0*runif(1)) + 1.1 ) ), times = iterations );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt:
```diff
@@ -0,0 +1,28 @@
+{{alias}}( x )
+    Computes the inverse hyperbolic cotangent of a number.
+
+    If `x >= -1 && x <= 1`, the function returns `NaN`.
```
```suggestion
    If provided a value on the interval [-1, 1], the function returns `NaN`.
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt:
```diff
@@ -0,0 +1,28 @@
+
+    Returns
+    -------
+    y: number|out
```
```suggestion
    y: number
```

BrunoFenzl replied:
I thought this was causing my linting error check because the function can return a number or NaN. I added it becasue the REPL documentation provides a case with multiple output values. Is `out` in this case the wrong return type or it should return just one type?

kgryte replied:
`NaN` is a number value, despite its "not-a-number" name. More specifically, it is a particular set of bit sequences which do not have a decimal representation when read from a value stored in double-precision floating-point format. Accordingly, even if a function returns `NaN`, using the `number` type is still appropriate.

Additionally, in this case, `out` is not a value type, so would not be appropriate in any context for parameter or return value types in REPL help documentation or in JSDoc type annotations.

You are correct, some functions in the project return values of different "types", but this is not one of those cases. :)

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/docs/repl.txt:
```diff
@@ -0,0 +1,28 @@
+    Examples
+    --------
+    > var y = {{alias}}( 1.0 )
+    0
```
```suggestion
    0.0
```

I believe, based on the documentation, that this should be `NaN`, not `0.0`.

BrunoFenzl replied:
NaN is for values in between -1 and 1. for these exact values, the javascript implementation return Infinity and I set this to 0 so it is more consistent with other functions.

kgryte replied:
For clarity, set to `0` to be consistent with what functions?

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js:
```diff
@@ -0,0 +1,96 @@
+/**
+* Computes the inverse hyperbolic cotangent of a number.
+*
+* ## Method
```
You can remove the method section in its entirety. This was only applicable for `acosh`, where we needed to explain in more detail how the implementation was determined. For `acoth`, the implementation is more straightforward.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js:
```diff
@@ -0,0 +1,96 @@
+		return 0;
+	}
+
+	return (ln((x + 1) / x) + ln(x / (x - 1))) / 2;
```
This implementation does not seem right to me. Is this still a work-in-progress?

BrunoFenzl replied:
I'm no mathematician so I hope I'm not utterly unfit the job! I derived this implementation from [wolframalpha.com](https://www.wolframalpha.com/input/?i=acoth(2)). One other way of writing this I found later in [wikipedia](https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions#Inverse_hyperbolic_cotangent) was `(1/2) * Math.log((x + 1) / (x - 1))`. I confess I didn't benchmark the difference.
For the case `x === 1.0 || x === -1.0` the function would return Infinity, so I thought 0 would be more consistent with other functions.

kgryte replied:
Ah. Okay. Thanks for the clarification.

As an aside, you don't have to be a mathematician to implement mathematical functions; however, I will say that it does take a while to get a "feel" for implementing numerical algorithms. But this is something which is not conferred by credentials, but more by practice.

As a first part of this journey, one thing we rarely do is implement textbook formulas. The reality is that numerical algorithms are often approximations which frequently optimize for particular value regimes. A perhaps more relevant example than `acosh` is [`atanh`](https://github.com/stdlib-js/stdlib/blob/07a705b6ce5aa3147951f737cdb8e4ee5817a991/lib/node_modules/%40stdlib/math/base/special/atanh/lib/atanh.js#L102). Notice how the Wikipedia formulas for both `atanh` and `acoth` are rather similar, but, for `atanh` (derived from FreeBSD), we are not implementing the textbook formula. Instead, we leverage `log1p` for better handling of small values and also circumvent computation entirely for tiny values.

In short, going back to the RFC, in this case, similar to, e.g., Julia, I would recommend that the implementation literally be

```javascript
var atanh = require( '@stdlib/math/base/special/atanh' );

function acoth( x ) {
    return atanh( 1.0/x );
}
```

This will hopefully resolve all the accuracy test failures in this package's unit tests and will obviate the need for any special value consideration. For example, since `atanh( 1.0 ) = +infinity`, the return value of `acoth( 1.0 )` will naturally be zero.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/lib/index.js:
```diff
@@ -0,0 +1,46 @@
+* var acoth = require( '@stdlib/math/base/special/acoth' );
+*
+* var v = acoth( 1.0 );
+* // returns 0
```
```suggestion
* // returns 0.0
```

However, I believe, based on the documentation, this should be `NaN`, not `0.0`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/package.json:
```diff
@@ -0,0 +1,71 @@
+    "stdmath",
+    "mathematics",
+    "math",
+    "math.acoth",
```
This keyword can be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/package.json:
```diff
@@ -0,0 +1,71 @@
+    "trigonometry",
+    "radians",
+    "angle",
+    "polyfill",
```
This can be removed, as this package does not serve as a polyfill, given that nothing at the standards level is happening in order to add built-in support for `acoth`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/package.json:
```diff
@@ -0,0 +1,71 @@
+    "radians",
+    "angle",
+    "polyfill",
+    "ponyfill"
```
This keyword can be removed for reasons similar to `polyfill`.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/scripts/accuracy.js:
```diff
@@ -0,0 +1,57 @@
+/**
```
This script can be removed.

kgryte replied:
This file still needs to be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env julia
```
This file can be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/REQUIRE:
```diff
@@ -0,0 +1,2 @@
+julia 0.5
```
This file can be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/scripts/fixtures/julia/data.json:
```diff
@@ -0,0 +1 @@
+{"expected":[0.0,0.3132037050764154,0.4411633162860367,0.5381861171312899,0.6190392084062233,0.6894729350848828,0.7524517545257728,0.8097457376945989,0.8625140249002963,0.911565758147417,0.9574929259372565,1.0007445375072996,1.0416710186716507,1.0805522693557368,1.1176161895584358,1.153051361233237,1.1870159963991522,1.2196444155578503,1.25105184338384,1.2813380281063398,1.310590019958354,1.33888 ...
```
This file can be removed.

kgryte replied:
This file still needs to be removed.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl:
````diff
@@ -0,0 +1,76 @@
+# Examples
+
+``` julia
+julia> x = linspace( -1000, 1000, 2001 );
````
```suggestion
julia> x = linspace( -1000.0, 1000.0, 2001 );
```

kgryte replied:
Need to add decimals here as shown in the suggestion.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,76 @@
+dir = dirname( file );
+
+# Positive medium values:
+x = linspace( 1.0, 3.0, 500 );
```
```suggestion
x = linspace( 1.0, 3.0, 503 );
```

Avoid nice even array lengths when generating fixtures. Otherwise, one generates predictable bit patterns, which may mask implementation flaws.

BrunoFenzl replied:
Nice! Good hint.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,76 @@
+gen( x, "medium_positive.json" );
+
+# Large positive values:
+x = linspace( 3.0, 28.0, 500 );
```
```suggestion
x = linspace( 3.0, 28.0, 503 );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,76 @@
+gen( x, "large_positive.json" );
+
+# Larger positive values:
+x = linspace( 28.0, 100.0, 500 );
```
```suggestion
x = linspace( 28.0, 100.0, 503 );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,76 @@
+gen( x, "larger_positive.json" );
+
+# Huge positive values:
+x = linspace( 1e300, 1e308, 1001 );
```
```suggestion
x = linspace( 1e300, 1e308, 1003 );
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/test/test.js:
```diff
@@ -0,0 +1,158 @@
+	t.end();
+});
+
+tape( 'the function returns `NaN` if provided value between `-1` and `1`', function test( t ) {
```
```suggestion
tape( 'the function returns `NaN` if provided a value between `-1` and `1`', function test( t ) {
```

Inclusive or exclusive? Not clear. The docs seem to suggest inclusive, but various example output suggests otherwise.

BrunoFenzl replied:
Should be exclusive.

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/test/test.js:
```diff
@@ -0,0 +1,158 @@
+	var i;
+
+	for ( i = -1; i < 1e3; i++ ) {
+		v = -(randu()*1.0e6) + (1-EPS);
```
```suggestion
		v = -(randu()*1.0e6) + (1.0-EPS);
```

We include decimals for code documentation purposes when a value and/or operations involve real-valued numbers. If a value can only be integer-valued, project convention is to omit the decimal.

BrunoFenzl replied:
ah! That's why... I was going to ask you why those units were being written with decimal points. Thanks for the clarification.

kgryte reviewed (changes requested):
@BrunoFenzl Thanks for working on this! This PR is useful for us, as it highlights the need for us to clarify what files are needed when, particularly when it comes to testing and benchmarking math implementations. For example, several of the files which are part of `acosh` are not applicable here. Sorry about this! Thanks for your patience and perseverance!

BrunoFenzl commented:
Wow! Thanks for the thoroughly review and sorry for the tons of inconsistencies! I'll take more care on the other functions I did, specially concerning the implementation in other languages. I have never seen such a complete testing suite (catching docs inconsistencies is awesome btw.) so I used the `acosh` folder as a template trying not to miss anything. I really hope I didn't take too much time from you.

kgryte commented:
@BrunoFenzl No worries! This is part of the process. The reality is that getting up-to-speed with all the project conventions in addition to implementing numerical algorithms can be a dizzying task (something the project needs to do a better job of addressing by better on-boarding and documentation). So thanks for putting in all the work!

BrunoFenzl commented:
@kgryte I think this PR is good now. Please let me know what I could improve if you find something wrong.

kgryte commented:
@BrunoFenzl Awesome! Will review shortly... :)

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/benchmark/python/benchmark.py:
```diff
@@ -0,0 +1,97 @@
+
+def benchmark():
+    """Run the benchmark and print benchmark results."""
+    setup = "from math import acoth; from random import random;"
```
```suggestion
    setup = "from math import atanh; from random import random;"
```

Python does not have an `acoth` function. You need to import `atanh` as this is the function used at L76.

BrunoFenzl replied:
strange, I had this fixed... something got missing. Will push it again

kgryte commented on lib/node_modules/@stdlib/math/base/special/acoth/lib/acoth.js:
```diff
@@ -0,0 +1,65 @@
+* // returns NaN
+*/
+function acoth( x ) {
+	if (x === 0.0) {
```
Why is this special case present? The inverse hyperbolic cotangent is not defined for `0.0`.

BrunoFenzl replied:
atanh should return 0.0 but is returning NaN. I probably missed something, but I could not figure it out why.

kgryte replied:
`atanh` **does** return `0.0`. To see this in action, start the stdlib REPL,

```
$ make repl
```

In the REPL,

```
> base.atanh( 0.0 )
0.0
```

For the inverse hyperbolic cotangent `acoth`, the equivalent `atanh` call is

```
> base.atanh( 1.0 / 0.0 )
NaN
```

This result is the same in Julia, Python, etc.

That the result should be `NaN` and not `0.0` can be confirmed by examining the plot of `acoth` where the domain is not defined in the region between (-1,1), as also stated in this package's documentation.

Hope this helps! :)

kgryte reviewed (changes requested):
@BrunoFenzl Did a quick review. Most of the remaining changes I flagged concern files which are not applicable to `acoth`.

kgryte commented:
@BrunoFenzl Just for reference, looking at the PR commit log, instead of merging from your fork into what I assume to be the cloned project repository, you want to merge from the project remote (i.e., this repo) into your fork. This will help prevent mixing your contributions with updates from the upstream repository (e.g., this PR indicates that `119` files have changed, when this PR really only concerns about `20` files). For example, assuming you have forked the project and cloned the fork to your local machine,

```
$ cd ./path/to/your/fork/stdlib
$ git checkout <your_feature_branch>
$ git remote -v
$ git remote add upstream https://github.com/stdlib-js/stdlib.git
$ git remote -v
$ git fetch upstream
$ git merge upstream/develop
```
I believe this is the workflow for syncing a feature development branch (e.g., `feature/acoth`) with changes in this repository.

Apologies if this is what you did! Something just seems a bit off with the commit log.

kgryte commented:
Ah...wait...did you commit directly to a `develop` branch in your fork? That seems to be why. Would you mind porting your changes to a `feature/*` branch (see also this [guide](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md#step-2-branch))? And then reopening a new PR from that feature branch to `develop` on this repository?

Sorry about this! I should have caught this earlier! In short, because this PR is attempting to go from `develop` to `develop`, the PR's commits are being interleaved with changes in the upstream repository, which will be problematic when I "squash-and-merge" this PR.

BrunoFenzl commented:
Oh! Sorry for the silly mistake! I did fetch upstream to synch with develop and forgot to switch back to my feature branch. I'll make a new pull request from a clean branch. BTW, I got rid of those errors in julia with version 0.5.2. It was no issue at all downgrading it. Since I'm on a mac, I just deleted the app, removed the preferences and installed again. Sorry for the extra work!

kgryte commented:
@BrunoFenzl No worries! We need to do a better job of making contributing a smoother process. Really appreciate your persistence and continued work!
